File: server/src/main/kotlin/starfield/routing/Game.kt
package starfield.routing

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.routing.*
import io.ktor.server.sessions.*
import java.util.Random
import kotlinx.serialization.Serializable
import starfield.*
import starfield.data.dao.CardDao
import starfield.engine.OracleId
import starfield.engine.Zone
import starfield.plugins.*

@Serializable
data class VirtualIdsMessage(
        val virtualIds: Map<Id, OracleId>,
        val oracleInfo: Map<OracleId, CardDao.OracleCard>
)

@Serializable
data class SideboardingVirtualIdsMessage(
        val main: Map<Id, OracleId>,
        val side: Map<Id, OracleId>,
        val oracleInfo: Map<OracleId, CardDao.OracleCard>
)

fun Route.engineRouting() {
    get("virtual-ids/library") {
        val session =
                call.sessions.get<UserSession>()
                        ?: return@get call.respondError(
                                "You must be logged in",
                                status = HttpStatusCode.Unauthorized
                        )

        val game = findBoardGame(session.id) ?: return@get call.respondError("Not in a game")
        val ids = game.assignVirtualIds(session.id)[Zone.LIBRARY]!!
        game.broadcastLogMessage(FindCardLogMessage(), session.id)
        val oracleInfo =
                CardDao().getCards(ids.values).associate {
                    val card = it.toOracleCard()
                    Pair(card.id, card)
                }

        call.respondSuccess(VirtualIdsMessage(ids, oracleInfo))
    }

    get("virtual-ids/sideboarding") {
        val session =
                call.sessions.get<UserSession>()
                        ?: return@get call.respondError(
                                "You must be logged in",
                                status = HttpStatusCode.Unauthorized
                        )

        val game = findBoardGame(session.id) ?: return@get call.respondError("Not in a game")
        val ids = game.assignVirtualIds(session.id, scoop = true)
        game.broadcastLogMessage(SideboardLogMessage(), session.id)
        val oracleInfo =
                CardDao().getCards(ids.values.flatMap { it.values }).associate {
                    val card = it.toOracleCard()
                    Pair(card.id, card)
                }

        val main =
                ids.entries
                        .filter { it.key != Zone.SIDEBOARD }
                        .map { it.value.entries }
                        .flatten()
                        .associate { Pair(it.key, it.value) }

        val side = ids[Zone.SIDEBOARD]!!

        call.respondSuccess(SideboardingVirtualIdsMessage(main, side, oracleInfo))
    }

    post("{id}/spectate") {
        val session =
                call.sessions.get<UserSession>()
                        ?: return@post call.respondError(
                                "You must be logged in",
                                status = HttpStatusCode.Unauthorized
                        )
        val preexistingGame = findGame(session.id)
        if (preexistingGame != null) {
            return@post call.respondError("Can't spectate while in another game")
        }
        val gameId = tryParseUuid(call.parameters["id"])
                        ?: return@post call.respondError("Invalid game id")
        val game = findAnyBoardGame(gameId)
                        ?: return@post call.respondError("Game not found", HttpStatusCode.NotFound)
        game.addSpectator(session.user())
        call.respondSuccess(true)
    }

    delete("{id}/spectate") {
        val session =
                call.sessions.get<UserSession>()
                        ?: return@delete call.respondError(
                                "You must be logged in",
                                status = HttpStatusCode.Unauthorized
                        )
        val gameId = tryParseUuid(call.parameters["id"])
            ?: return@delete call.respondError("Invalid game id")
        val game = findGameSpectating(session.id)
            ?: return@delete call.respondError("Game not found", HttpStatusCode.NotFound)
        if (game.id != gameId) {
            return@delete call.respondError("Not spectating this game")
        }
        game.removeSpectator(session.id)
        call.respondSuccess(true)
    }

    get("search/tokens") {
        val dao = CardDao()
        val name = call.parameters["name"]
        val type = call.parameters["type"]
        val pt = call.parameters["pt"]
        val text = call.parameters["text"]
        val color = call.parameters["colors"]

        val types = mutableListOf<String>()

        if (type != null) {
            val rawTypes = type.split(" - ")
            if (rawTypes.size == 2) {
                types.addAll(rawTypes[1].split(" "))
                types.addAll(rawTypes[0].split(" "))
            } else {
                types.addAll(rawTypes[0].split(" "))
            }
        }

        val tokens = dao.searchForTokens(name, color, types, text, pt)
        call.respondSuccess(tokens.map { it.toOracleCard() })
    }

    get("search/cards") {
        val dao = CardDao()
        val name = call.parameters["name"] ?: return@get call.respondError("Must specify a name")
        val cards = dao.searchForCards(name)
        call.respondSuccess(cards.map { it.toOracleCard() })
    }

    get("roll-die") {
        val session =
                call.sessions.get<UserSession>()
                        ?: return@get call.respondError(
                                "You must be logged in",
                                status = HttpStatusCode.Unauthorized
                        )
        val game = findBoardGame(session.id) ?: return@get call.respondError("Not in a game")

        val sides =
                call.parameters["sides"]?.toIntOrNull()
                        ?: return@get call.respondError("Must specify sides")

        val rng = Random()
        val result = rng.nextInt(sides) + 1

        game.broadcastLogMessage(RollDieLogMessage(sides, result), session.id)

        return@get call.respondSuccess(result)
    }
}

File: client/src/api/game.ts
import type { OracleId } from "@/stores/board";
import { deleteJson, deleteJsonRaw, getJson, postJson } from ".";
import type { OracleCard } from "./message";

type OracleIdMap = { [key: string]: OracleId };
type VirtualIdsMessage = {
  virtualIds: OracleIdMap,
  oracleInfo: { [oracleId: OracleId]: OracleCard }
}
type SideboardingVirtualIdsMessage = {
  main: OracleIdMap,
  side: OracleIdMap,
  oracleInfo: { [oracleId: OracleId]: OracleCard }
}

export function getVirtualIds(): Promise<VirtualIdsMessage> {
  return getJson('game/virtual-ids/library');
}

export function getVirtualScryIds(count: number): Promise<VirtualIdsMessage> {
  return getJson(`game/virtual-ids/scry?count=${count}`);
}

export function getVirtualSideboardingIds(): Promise<SideboardingVirtualIdsMessage> {
  return getJson('game/virtual-ids/sideboarding');
}

export function rollDie(sides: number): Promise<number> {
  return getJson(`game/roll-die?sides=${sides}`);
}

export function beginSpectating(gameId: string): Promise<void> {
  return postJson(`game/${gameId}/spectate`, {});
}

export async function stopSpectating(gameId: string): Promise<boolean> {
  const result = await deleteJsonRaw(`game/${gameId}/spectate`);
  return result.success;
}

export function searchForTokens(name?: string, type?: string, colors?: string, text?: string, pt?: string): Promise<OracleCard[]> {
  const params = { name, type, colors, text, pt };
  
  const searchParams = new URLSearchParams();
  for(const [key, value] of Object.entries(params)) {
    if(value) {
      searchParams.append(key, value);
    }
  }

  return getJson(`game/search/tokens?${searchParams.toString()}`);
}

export function searchForCards(name: string): Promise<OracleCard[]> {
  return getJson(`game/search/cards?name=${name}`);
}